sound/soc/sof/ipc4-topology.c

```diff
@@ -0,0 +1,633 @@
+		goto finish;
+	}
+	dev_dbg(scomp->dev, "dai %s gtw_cfg dma_buffer_size %u\n",
+		swidget->widget->name, ipc4_dai->copier.gtw_cfg.dma_buffer_size);
+
+	ret = sof_update_ipc_object(scomp, &node_type,
+				    SOF_IPC4_COPIER_TOKENS, swidget->tuples,
+				    swidget->num_tuples, sizeof(node_type), 1);
+
+	if (ret != 0) {
+		dev_err(scomp->dev, "parse dai copier_node tokens failed %d\n",
+			ret);
+		goto finish;
+	}
+	dev_dbg(scomp->dev, "dai %s node_type %u\n",
+		swidget->widget->name, node_type);
+
+	ipc4_dai->copier.gtw_cfg.node_id = SOF_IPC4_NODE_TYPE(node_type);
+
+	dai->scomp = scomp;
+
+	dev_dbg(scomp->dev, "dai %s cpc %d\n", swidget->widget->name,
+		ipc4_dai->copier.base_config.cpc);
+
+finish:
+	dai->private = ipc4_dai;
+	return ret;
+}
+
+static int sof_ipc4_widget_setup_comp_pipeline(struct snd_sof_widget *swidget)
+{
+	struct snd_soc_component *scomp = swidget->scomp;
+	struct sof_ipc4_pipeline *pipeline;
+	int ret;
+
+	pipeline = kzalloc(sizeof(*pipeline), GFP_KERNEL);
+	if (!pipeline)
+		return -ENOMEM;
+
+	ret = sof_update_ipc_object(scomp, pipeline, SOF_SCHED_TOKENS, swidget->tuples,
+				    swidget->num_tuples, sizeof(*pipeline), 1);
+	if (ret != 0) {
+		dev_err(scomp->dev, "parse pipeline tokens failed\n");
+		goto err;
+	}
+
+	/* TODO: Get priority from topology */
+	pipeline->priority = 0;
+
+	dev_dbg(scomp->dev, "pipeline %s: id %d pri %d lp mode %d\n",
+		swidget->widget->name, swidget->pipeline_id,
+		pipeline->priority, pipeline->lp_mode);
+
+	swidget->private = pipeline;
+
+	return 0;
+err:
+	kfree(pipeline);
+	return ret;
+}
+
+/*
+ * PGA Topology
+ */
+
+static int sof_ipc4_widget_setup_comp_pga(struct snd_sof_widget *swidget)
+{
+	struct snd_soc_component *scomp = swidget->scomp;
+	struct sof_ipc4_gain *gain;
+	int ret;
+
+	gain = kzalloc(sizeof(*gain), GFP_KERNEL);
+	if (!gain)
+		return -ENOMEM;
+
+	swidget->private = gain;
+
+	/*
+	 * available_fmt->out_audio_fmt will be ignored since there is no
+	 * out_audio_fmt in gain.
+	 */
+	ret = sof_ipc4_get_audio_fmt(scomp, swidget, &gain->available_fmt, false);
+	if (ret != 0)
+		goto err;
+
+	ret = sof_update_ipc_object(scomp, &gain->data, SOF_GAIN_TOKENS, swidget->tuples,
+				    swidget->num_tuples, sizeof(gain->data), 1);
+	if (ret != 0) {
+		dev_err(scomp->dev, "parse gain tokens failed\n");
+		goto err;
+	}
+
+	dev_dbg(scomp->dev, "tplg2: ready widget %s, ramp_type %d, duration %d, val %d  cpc %d",
+		swidget->widget->name, gain->data.curve_type, gain->data.curve_duration,
+		gain->data.init_val, gain->base_config.cpc);
+
+	return 0;
+err:
+	kfree(gain);
+	return ret;
+}
+
+static int sof_ipc4_widget_setup_comp_mixer(struct snd_sof_widget *swidget)
+{
+	struct snd_soc_component *scomp = swidget->scomp;
+	struct sof_ipc4_mixer *mixer;
+	int ret;
+
+	mixer = kzalloc(sizeof(*mixer), GFP_KERNEL);
+	if (!mixer)
+		return -ENOMEM;
+
+	swidget->private = mixer;
+
+	/*
+	 * available_fmt->out_audio_fmt will be ignored since there is no
+	 * out_audio_fmt in gain.
```

Comment:
this comment is ambiguous. What you mean to say is "The out_audio_fmt in topology is ignored as it is not required to be sent to the FW" right?

Reply:
> this comment is ambiguous. What you mean to say is "The out_audio_fmt in topology is ignored as it is not required to be sent to the FW" right?

Yes, thanks.